cookbook/controller/upload_file.rst

```diff
@@ -164,15 +188,220 @@ There are some important things to consider in the code of the above controller:
    that information. That's why it's always better to generate a unique name and
    use the :method:`Symfony\\Component\\HttpFoundation\\File\\UploadedFile::guessExtension`
    method to let Symfony guess the right extension according to the file MIME type;
-#. The ``UploadedFile`` class also provides a :method:`Symfony\\Component\\HttpFoundation\\File\\UploadedFile::move`
-   method to store the file in its intended directory. Defining this directory
-   path as an application configuration option is considered a good practice that
-   simplifies the code: ``$this->container->getParameter('brochures_dir')``.
 
-You can now use the following code to link to the PDF brochure of an product:
+You can use the following code to link to the PDF brochure of an product:
+
+.. configuration-block::
+
+    .. code-block:: html+twig
+
+        <a href="{{ asset('uploads/brochures/' ~ product.brochure) }}">View brochure (PDF)</a>
+
+    .. code-block:: html+php
+
+        <a href="<?php echo $view['assets']->getUrl('uploads/brochures/'.$product->getBrochure()) ?>">
+            View brochure (PDF)
+        </a>
+
+Creating an Uploader Service
+----------------------------
+
+To avoid logic in controllers, making them big, you can extract the upload
+logic to a seperate service::
+
+    // src/AppBundle/FileUploader.php
+    namespace AppBundle\FileUploader;
+
+    use Symfony\Component\HttpFoundation\File\UploadedFile;
+
+    class FileUploader
+    {
+        private $targetDir;
+
+        public function __construct($targetDir)
+        {
+            $this->targetDir = $targetDir;
+        }
+
+        public function upload(UploadedFile $file)
+        {
+            $fileName = md5(uniqid()).'.'.$file->guessExtension();
+
+            $file->move($this->targetDir, $fileName);
+
+            return $fileName;
+        }
+    }
+
+This class can be registered as a service in the service container:
```

Comment:
`This class can be registered as a service in the service container:`

->

`Then, define a service for this class:` ?